File: server/src/main/com/broll/gainea/server/core/bot/BotOptionalAction.java
package com.broll.gainea.server.core.bot;

import com.broll.gainea.net.NT_Action;

public abstract class BotOptionalAction<A extends NT_Action, O extends BotOptionalAction.BotOption> extends BotAction<A> {

    private O selectedOption;

    public abstract O score(A action);

    public void setSelectedOption(O selectedOption) {
        this.selectedOption = selectedOption;
    }

    public O getSelectedOption() {
        return selectedOption;
    }

    public static class BotOption {
        private float score;

        public BotOption(float score) {
            this.score = score;
        }

        public float getScore() {
            return score;
        }
    }

}

File: server/src/main/com/broll/gainea/server/core/bot/BotOptionalAction.kt
package com.broll.gainea.server.core.bot

import com.broll.gainea.net.NT_Action

abstract class BotOptionalAction<A : NT_Action, O : BotOptionalAction.BotOption> : BotAction<A>() {
    var selectedOption: O? = null
        private set

    abstract fun score(action: A): O?
    fun setSelectedOption(selectedOption: O) {
        this.selectedOption = selectedOption
    }

    open class BotOption(val score: Float)
}
